torch/csrc/distributed/c10d/ProcessGroup.hpp

```diff
@@ -105,173 +111,308 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
   virtual void endCoalescing(
       c10::DeviceType deviceType,
       std::vector<c10::intrusive_ptr<Work>>& reqs) {
-    // only nccl has implemented startCoalescing so only execute for nccl backends
+    // only nccl has implemented startCoalescing so only execute for nccl
+    // backends
     if (getBackendType() == BackendType::NCCL) {
       getBackend(deviceType)->endCoalescing(reqs);
     }
   }
 
-  // Consider using ops in Ops.hpp instead of the below, which route things
-  // to the dispatcher.
-  // TODO: Find a way to force the above rule programmatically.
   virtual c10::intrusive_ptr<Work> broadcast(
-      std::vector<at::Tensor>& /* tensors */,
-      const BroadcastOptions& /* opts */ = BroadcastOptions()) {
-    TORCH_CHECK(
-        false,
-        c10::str(
-            "ProcessGroup ", getBackendName(), " does not support broadcast"));
+      std::vector<at::Tensor>& tensors,
+      const BroadcastOptions& opts = BroadcastOptions()) {
+    static auto op =
+        c10::Dispatcher::singleton()
+            .findSchemaOrThrow("c10d::broadcast_", "")
+            .typed<
+                std::tuple<std::vector<at::Tensor>, c10::intrusive_ptr<Work>>(
+                    at::TensorList,
+                    const c10::intrusive_ptr<::c10d::ProcessGroup>&,
+                    int64_t,
+                    int64_t,
+                    int64_t)>();
+    // It's awakward to unbox the opts here and box them again in the custom C++
+    // op. But it's also complicated to make opts as a CustomClassHolder. Leave
+    // it as it is now.
+    return std::get<1>(op.call(
+        tensors,
+        c10::intrusive_ptr<ProcessGroup>::unsafe_reclaim_from_nonowning(this),
+        opts.rootRank,
+        opts.rootTensor,
+        opts.timeout.count()));
```

Comment:
I notice that we still have a copy in Ops.cpp. I wonder if we could call `ops::broadcast` here?

Reply:
I tried that but unfortunately `Ops.hpp` also includes `ProcessGroup.hpp` so there becomes a circular dependency. Probably the follow up work is to remove the definitions in Ops.cpp (after we also update reducer.cpp)